images/echoserver/echoserver.go

```diff
@@ -121,3 +161,63 @@ func processError(w http.ResponseWriter, err error, code int) {
 	w.WriteHeader(code)
 	w.Write(body)
 }
+
+func listenAndServeTLS(addr string, serverCert string, serverPrivKey string, clientCA string, handler http.Handler) error {
+	var config tls.Config
+
+	// optionally enable client certificate validation if client CA certificates are given
+	if clientCA != "" {
+		ca, err := ioutil.ReadFile(clientCA)
+		if err != nil {
+			return err
+		}
+
+		certPool := x509.NewCertPool()
+		if ok := certPool.AppendCertsFromPEM(ca); !ok {
+			return fmt.Errorf("unable to append certificate in %s to CA pool", clientCA)
+		}
+
+		config.ClientAuth = tls.RequireAndVerifyClientCert
```

Comment:
What's the rationale for using `tls.RequireAndVerifyClientCert` here? For a test server, I'd suggest that it makes sense to request a client certificate (and report whether it was sent), but not fail if the peer doesn't send one. Some test scenarios will need to send a certificate and others won't.

Reply:
I changed it to `tls.VerifyClientCertIfGiven` to allow unauthenticated clients, while still requesting and validating client certificates (if given). Connection will be still rejected if given certificates cannot be validated, making it easy to recognize in test case that authentication was done successfully.